File: one_dimension/general_euler.r
General_Euler <- function(num_of_paths, num_of_steps, mu, sigma, Texp=1, X0=0, convert_y_to_x=function(y) y, convert_x_to_y=function(x) x){
    # Simulate a 1-dimensional process using Monte Carlo with the standard Euler scheme.
    # This can be used to simulate the payoff of an option and calculate the implied volatility smile.
    #
    # Args:
    #     num_of_paths (integer): The number of Monte Carlo paths
    #     num_of_steps (integer): The number of steps in each Monte Carlo path
    #     mu (function): The drift coefficient, a function of t (numeric) and X (array-like) returning a scalar
    #     sigma (function): The diffusion coefficient, a function of t (numeric) and X (array-like) returning a scalar
    #     Texp (Optional[numeric]): Time, in years, to simulate process.  Defaults to 1.
    #     X0 (Optional[numeric]): Initial value for the estimator process, X. Defaults to 0.
    #     convert_y_to_x (Optional[function]): A function to convert Y_t to X_t. Defaults to X_t <- Y_t
    #     convert_x_to_y (Optional[function]): A function to convert X_t to Y_t. Defaults to Y_t <- X_t
    # Raises:
    #     Error: If mu is not a function.
    #     Error: If sigma is not a function.
    # Returns:
    #     (list of function(s)):  Currently returns a list holding a function which runs a simulation
    if (!is.function(mu)) stop('mu must be a function')
    if (!is.function(sigma)) stop('sigma must be a function')

    dt <- Texp / num_of_steps
    Y0 <- convert_x_to_y(X0)

    run_monte_carlo <- function(g_maker, strikes, num_paths_to_plot=FALSE){
        # Runs the Monte Carlo simulation to calculate E[g(X_T)] for different strikes.
        #
        # Args:
        #     g_maker (function): A function that returns a function f(x) <- g(x,K)
        #     strikes (array-like): An array of strikes to pass in to g_maker
        #     num_paths_to_plot (Optional[integer or logical]): How many paths should be plotted. Defaults to FALSE (no plot). If TRUE, defaults to 100.
        # Returns:
        #     (array-like): A array of E[g(X_T)] for the range of strikes

        # Simulates the process X_t across time.
        # X is a 2-dimensional matrix with each row representing a path and each entry within it a single timestep.

        dW <- matrix(rnorm(num_of_paths*num_of_steps,sd=sqrt(dt)), nrow=num_of_paths, ncol=num_of_steps)

        # Generate the process Y using the Lamperti transform version of the process X
        
        # Initialize Y0. More efficient to just initialize all values as Y0
        Y <- matrix(Y0, nrow=num_of_paths, ncol=num_of_steps+1)

        # Step through time
        for (i in 1:num_of_steps){
            t <- (i-1)*dt
            Y[,i+1] <- Y[,i] + mu(t, Y[,i]) * dt + sigma(t, Y[,i]) * dW[,i]
        }

        # Plotting
        if (num_paths_to_plot != FALSE){ # Plots the first num_paths_to_plot paths.
            if (num_paths_to_plot == TRUE) num_paths_to_plot <- 100 # Default value
            num_paths_to_plot <- min(num_paths_to_plot, num_of_paths)
            #matplot plots columns so we transpose
            matplot(seq(0, Texp, Texp/num_of_steps), t(convert_y_to_x(Y[1:num_paths_to_plot,])), main=sprintf('Generated paths (first %i paths) - Euler scheme',num_paths_to_plot), xlab='t', ylab='X_t', type='l', lty=1)
        }
        
        # Final prices for each path (Undo the Lamperti transform)
        X_Ts <- convert_y_to_x(Y[,ncol(Y)])

        # Returns the mean of psi across all paths for each strike
        # return(vapply(strikes, function(K) mean(g_maker(K)(X_Ts)), FUN.VALUE=1))

        # Calculates prices for each path by row, each strike by column
        price_matrix <- vapply(strikes, function(K) g_maker(K)(X_Ts), FUN.VALUE=numeric(num_of_paths))
        print(data.frame(K=strikes, Variance=diag(var(price_matrix)))) # Diagonals of cov mtx are column variances
        return(colMeans(price_matrix))
    }
    
    # Return a list holding the function so we can call it object oriented style
    return(list(run_monte_carlo=run_monte_carlo))
}

File: one_dimension/general_euler.py
import numpy as np
import matplotlib.pyplot as plt

class General_Euler:
    """
    Simulate a 1-dimensional process using Monte Carlo with the standard Euler scheme.
    This can be used to simulate the payoff of an option and calculate the implied volatility smile.
    """
    def __init__(self, num_of_paths, num_of_steps, mu, sigma, T=1, X0=0):
        """
        Args:
            num_of_paths (int): The number of Monte Carlo paths
            num_of_steps (int): The number of steps in each Monte Carlo path
            mu (function): The drift coefficient, a function of t (float) and X (np.array) returning a scalar
            sigma (function): The diffusion coefficient, a function of t (float) and X (np.array) returning a scalar
            T (Optional[float]): Time, in years, to simulate process.  Defaults to 1.
            X0 (Optional[float]): Initial value for the estimator process, X. Defaults to 0.
        Raises:
            TypeError: If mu is not a function.
            TypeError: If sigma is not a function.
        """
        if not hasattr(mu, '__call__'): raise TypeError('mu must be a function')
        if not hasattr(sigma, '__call__'): raise TypeError('sigma must be a function')
        
        self.num_of_paths = num_of_paths
        self.num_of_steps = num_of_steps
        self.mu = mu
        self.sigma = sigma
        self.T = float(T)
        self.X0 = float(X0)

    def simulate_process(self):
        """
        Simulates the process self.X across time.
        self.X is a 2-dimensional np.array with each row representing a single path.
        Each entry in the row represents a single timestep.
        """
        dt = self.T / self.num_of_steps
        self.dW = np.sqrt(dt)*np.random.normal(size=(self.num_of_paths, self.num_of_steps))
        self.X = np.empty((self.num_of_paths, self.num_of_steps+1))
        self.X[:,0] = self.X0
        for t in xrange(self.num_of_steps):
            self.X[:,t+1] = self.X[:,t] + self.mu(t*dt, self.X[:,t]) * dt + self.sigma(t*dt, self.X[:,t]) * self.dW[:,t]
    
    def run_monte_carlo(self, g_maker, strikes, plot_paths=False, convert_y_to_x_func=lambda y: y):
        """
        Runs the Monte Carlo simulation to calculate E[g(X_T)] for different strikes.
        
        Args:
            g_maker (funcion): A function that returns a function f(x) = g(x,K)
            strikes (np.array): An array of strikes to pass in to g_maker
            plot_paths (Optional[boolean]): Should the paths be plotted. Defaults to False
            convert_y_to_x_func (Optional[function]): A function to convert Y_T to X_T. Defaults to X_T = Y_T
        Returns:
            (np.array): A array of E[g(X_T)] for the range of strikes
        """
        self.simulate_process()
        if plot_paths: self.plot_paths()
        
        X_Ts = [convert_y_to_x_func(X_T) for X_T in self.X[:,-1]]
        
        return np.array([np.mean(g_maker(K)(X_Ts)) for K in strikes])

    def plot_paths(self, paths_to_plot=100):
        """
        Plots the first paths_to_plot paths.
        """
        paths_to_plot = min(paths_to_plot, len(self.X))
        for path in self.X[:paths_to_plot]:
            plt.plot(path)
        plt.title('Generated paths (first {0} paths) - Euler scheme'.format(paths_to_plot))
        plt.xlabel('t')
        plt.ylabel('X_t')
        plt.show()
